dashboard/lib/dynamic_config/datastore_cache.rb

```diff
@@ -0,0 +1,93 @@
+# A caching layer that sits in front of a datastore that
+# implements get and set
+class DatastoreCache
+
+  class ConcurrentRefreshError < StandardError
+  end
+
+  # @param datastore [Object]
+  # @param cache_expiration [int] seconds after which a cached entry expires
+  def initialize(datastore, cache_expiration: 30)
+    @cache = {}
+    @datastore = datastore
+    @cache_expiration = cache_expiration
+
+    @global_lock = Mutex.new
+    @per_key_locks = {}
+
+    seed_cache
+  end
+
+  # Gets the data associated with a given key
+  # @param key [String]
+  # @returns [?] stored value
+  def get(key)
+    if @cache.key? key
+      cached_value, inserted_at = @cache[key]
+      if Time.now.to_f - inserted_at < @cache_expiration
+        return cached_value
+      end
+    end
+
+    begin
+      return refresh(key)
+    rescue
+      return @cache[key][0] if @cache.key? key
+    end
+    nil
+  end
+
+  # Update the cached value from the datastore. This uses Mutex's to
+  # only allow one update per key at on time.
+  # @param key [String]
+  # @return value for key
+  def refresh(key)
```

Comment:
Please make refresh private.  (The cache is responsible for managing its own refreshing and callers should not call this directly.)